### PR TITLE
Split streaming server from web server

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This is a more recent attempt at #21143.

I've opted to not use Yarn Workspaces for this split, as in testing the functionality did not work in a [consistent and documented manner](https://discord.com/channels/231908446830723072/1105529545685217391), so instead I've opted for two completely separate projects with isolated dependencies. 

For ease of end-user use, these two projects are linked together via a `postinstall`: this means that they can still just do `yarn install && yarn start` to start the streaming server, without needed to change to: `cd streaming && yarn install && yarn start`.

As for CI/CD, I've moved the pipelines to use a matrix which enables building each dockerfile and using names for them in the github actions output.

This branch was last sync'd with `main` at 3869e8c21069a9bc927e93c904e00d68b004a5be.

I've also fixed a minor issue in the streaming server where it was attempting to load `.env` when running from a docker container; this could've lead to unpredictable behaviour: in docker it's always better to pass the environment via variables rather than loading a `.env` file from within the docker image.

For now, this doesn't change the main Dockerfile, which could be changed in the future to not ship node.js & the streaming server inside of it.

### CODE REVIEW

**Review by renchap — COMMENTED** [review]
Looks like a good start.

The various Github actions will need to be modified to build and push an image for the streaming server, named `mastodon-streaming` maybe?

File: package.json
Comment (ThisIsMissEm) [reply package.json:13]: I've left this as just start for legacy, next release after this we should remove it — just didn't want to break things too quickly

File: streaming/Dockerfile
Comment (ThisIsMissEm) [reply streaming/Dockerfile:2]: Yeah, I'm being very conservative with deviations from what we currently have, as the streaming server isn't tested, so I can't have confidence a change doesn't break things.

I'm currently toying with how we could test; I think injecting data to redis might be best, and then checking a client connection behaves as expected, but I'd like to try to more formally define how the real-time functionality should work, there's a lot of room for improvement, and we'll need a v2 eventually I think

File: streaming/package.json
Comment (nschonni) [inline streaming/package.json:2]: Since it's being split out to a org namespace, would it make sense to just move it out of this repo, and then reference the package or Dockerfile in this repo

Comment (ThisIsMissEm) [reply streaming/package.json:2]: For now, I don't think so. It's a private package not intended to be published. Additionally, how the streaming server works is explicitly coupled to how the rest of Mastodon works, and the contract between them isn't (yet) strictly defined.

File: streaming/package.json
Comment (ThisIsMissEm) [inline streaming/package.json:20]: One of the connection dependencies received an update, looks like I just need to regenerate the lockfile

**Comment by ThisIsMissEm** [comment]
(there isn't currently an issue on JSON Linting / lint (pull_request), I just got rate limited by the GitHub API apparently)

File: .github/workflows/build-image.yml
Comment (ThisIsMissEm) [inline .github/workflows/build-image.yml:24]: @renchap apparently the way to build multiple images is to use a matrix, idk how I feel about this / what this change could potentially break. (e.g., required actions)

File: streaming/package.json
Comment (ThisIsMissEm) [inline streaming/package.json:35]: These may not be needed, but I'm playing it safe by including them here, such that we can move the streaming server towards typescript in the future (I've seen some issues where missing these in mono-repos can result in them not being found by tooling and editors, despite the fact that they should be being inherited from the root package.json)

**Comment by ThisIsMissEm** [comment]
@Gargron @renchap I've resolved the merge conflicts here against `main` so hopefully we can get this merged.

**Comment by ThisIsMissEm** [comment]
@renchap I originally had this PR updating the dependabot config to include the `streaming/Dockerfile` and `streaming/package.json` / `streaming/yarn.lock` files, based on my reading of the Renovate documentation, those should be automatically matched now, without additional configuration. Is this a correct understanding?

File: package.json
Comment (ThisIsMissEm) [inline package.json:24]: @Gargron if you want, we can keep this, or we can get rid of it — the benefit of keeping it until the next major is that we wouldn't break how people upgrade their servers, they can still do a `yarn install` and get everything they need installed, despite under the hood the package.json files have been split.

I see this as a stop-gap until we can probably define the boundaries/protocols between rails and streaming, and ultimately rewrite streaming in Go, which would be done as a major-release change.

File: streaming/package.json
Comment (ThisIsMissEm) [inline streaming/package.json:16]: This package has a new major out, but it's gone ESM only, so we cannot upgrade it until streaming is ESM only. @renchap we may need to add this to the renovate configuration.

https://github.com/sindresorhus/is-docker/releases/tag/v3.0.0

File: streaming/index.js
Comment (ClearlyClaire) [inline streaming/index.js:25]: I'm not sure why we'd special-case this, and it seems unrelated to the split. Did I miss something?

Comment (ThisIsMissEm) [reply streaming/index.js:25]: Inside a docker container, the .env file doesn't need to be / shouldn't be losded, as all environment variables should come from the container runtime

Comment (ClearlyClaire) [reply streaming/index.js:25]: Yeah I get that, but that seems unrelated to the split, and I don't think the Rails part does that special-casing, so I'd omit this for simplicity and consistency, unless there is a stronger reason.

Comment (renchap) [reply streaming/index.js:25]: Also dotenv should use the environment variables by default if they are set, I am not sure it is an issue if it uses the defaults in `.env` / `.env.production`?

**Comment by renchap** [comment]
We would prefer to merge this after 4.2.0 is released, so reviewing it is on hold for now but the team is ok with the general idea of this PR and will review it once 4.2.0 has been cut off.

**Comment by ThisIsMissEm** [comment]
@renchap I think both this and #21143 should probably be closed, as they'll require too significant amount of work to rebase, and instead we just make sure doing this is tracked in an issue and planned accordingly.

**Comment by ThisIsMissEm** [comment]
Per conversation on [discord about this](https://discord.com/channels/231908446830723072/231908446830723072/1171939682989375700): we're now ready to proceed forwards with this change, however, in two phases:

1) Perform the split using yarn workspaces (I'll drop the dotenv change too)
2) A separate PR tackling the dockerfile and CI changes for publishing a separate docker image.

File: package.json
Comment (ThisIsMissEm) [inline package.json:4]: I've just moved this up, such that it's near the `engines` field and not forgotten about below the dependencies

**Comment by ThisIsMissEm** [comment]
I'm not sure if I should also change the root `package.json` to have `start` be `cd streaming && yarn start` such that the script path is only defined once.. it does mean that `yarn start` from root would technically be running yarn within yarn, but we do that elsewhere too.

Screenshot 2023-11-10 at 6 43 21 pm

**Comment by vmstan** [comment]
I have this running a dedicated streaming build based off a modified version of #26850 on vmst.io right now.

The image size is reduced from 835MB to 299MB.

**Comment by ThisIsMissEm** [comment]
@vmstan I also didn't do `yarn workspaces focus @mastodon/streaming` in the root dockerfile here, which should improve things even more.

**Comment by vmstan** [comment]
> @vmstan I also didn't do `yarn workspaces focus @mastodon/streaming` in the root dockerfile here, which should improve things even more.

That's how I have it running in my test.

File: Procfile.dev
Comment (mjankowski) [inline Procfile.dev:3]: If this is in the workspace, I think it can be something like `yarn workspace streaming start` ?

File: streaming/index.js
Comment (mjankowski) [inline streaming/index.js:21]: Is it possible to just pick one of these ways to run it, instead of handling either? ... or are there scenarios where we want to be running it different ways?

File: Procfile.dev
Comment (ThisIsMissEm) [reply Procfile.dev:3]: oooh, good call!

```suggestion
stream: env PORT=4000 yarn workspace @mastodon/streaming start
```

File: streaming/index.js
Comment (ThisIsMissEm) [reply streaming/index.js:21]: basically if someone is in `mastodon/` and runs `node streaming/index.js` then it'll use `mastodon/.env` or `mastodon/.env.production`, but if you run from `mastodon/streaming` then it'll attempt to use `mastodon/streaming/.env` or `mastodon/streaming/.env.production` and silently fails (you get an database error)

Comment (mjankowski) [reply streaming/index.js:21]: Fair enough. Is a pattern like this relevant? https://github.com/motdotla/dotenv/issues/238#issuecomment-348017037

Comment (ThisIsMissEm) [reply streaming/index.js:21]: That'd pull in two additional files, and also search for .env in the user's home directory and other places: https://www.npmjs.com/package/find-config?activeTab=readme

File: streaming/index.js
Comment (renchap) [inline streaming/index.js:27]: Could we rather load it with a relative path from `index.js`? This seems cleaner

```
path: `${__dirname}/../${dotenvFile}`
```

File: streaming/package.json
Comment (renchap) [inline streaming/package.json:15]: In the future we might want to have more script here, to build / typecheck / lint / … this part of the project independently. But at the moment, the root commands already cover `streaming/`, and I am not sure how to set it up properly to have them separate yet.

**Review by renchap — COMMENTED** [review]
Looks good to me overall, please see my comment about the env file location

File: streaming/index.js
Comment (ThisIsMissEm) [reply streaming/index.js:27]: You could, but that'd pose problems if we were to migrate to ESM, as `__dirname` isn't easily available in ESM, additionally, you don't want to use `/../` as that could break non-linux systems (that use a different path separator).

File: streaming/package.json
Comment (ThisIsMissEm) [reply streaming/package.json:15]: yeah, so, the root command could be now changed to `yarn workspace @mastodon/streaming start` instead of `cd streaming && yarn start`

Comment (ThisIsMissEm) [reply streaming/package.json:15]: But we should also consider `yarn start` at the root largely deprecated in favor of `yarn workspace @mastodon/streaming start`, but we need the shim in there for people who haven't migrated.

**Comment by ThisIsMissEm** [comment]
@renchap I have just noticed that the typechecking is a little broken, I'm fixing that in a future PR were I discovered it (like, it doesn't "break" the pipeline, but in my editor, the eslint/tsconfig configuration definitely breaks)

File: streaming/index.js
Comment (renchap) [reply streaming/index.js:27]: If we manage to switch to ESM, we can still handle this properly using `new URL(`../${dotenvFile}`, import.meta.url)`

For my example above, you can use `path.join` or `path.resolve` to work around it

Comment (renchap) [reply streaming/index.js:27]: Also this implementation is less error-prone, for example if the user launches the streaming process without the correct cwd (like it could happen when launched by systemd).
Referencing the `.env` file using the source-code's path is consistent with the way it is done in Ruby (`dotenv-ruby` uses `Rails.root` for this)

**Review by renchap — DISMISSED** [review]
👍

**Comment by renchap** [comment]
Can you rebase it, then we will merge it 👍

**Review by renchap — APPROVED** [review]

**Pull request merged**
